src/assets/ManifestAsset.js

```diff
@@ -0,0 +1,168 @@
+const path = require('path');
+const Asset = require('../Asset');
+
+/**
+ * A shared asset that handles:
+ * - PWA .webmanifest
+ * - PWA manifest.json
+ * - WebExtension manifest.json
+ */
+class ManifestAsset extends Asset {
+  constructor(name, options) {
+    super(name, options);
+    this.type =
+      path.basename(name) === 'manifest.json' ? 'json' : 'webmanifest';
+    this.isAstDirty = false;
+  }
+
+  parse(code) {
+    return JSON.parse(code);
+  }
+
+  processSingleDependency(path, opts) {
+    opts = opts || {entry: true};
+    return this.addURLDependency(path, opts);
+  }
+
+  processMultipleDependencies(filenames, opts) {
+    return filenames.map(filename =>
+      this.processSingleDependency(filename, opts)
+    );
+  }
+
+  processBackground(nodeName) {
+    if (nodeName !== 'background') {
+      return;
+    }
+
+    const background = this.ast[nodeName];
+    if (Array.isArray(background.scripts)) {
+      background.scripts = this.processMultipleDependencies(background.scripts);
+      this.isAstDirty = true;
+    }
+    if (background.page) {
+      background.page = this.processSingleDependency(background.page);
+      this.isAstDirty = true;
+    }
+  }
+
+  processContentScripts(nodeName) {
+    if (nodeName !== 'content_scripts') {
+      return;
+    }
+
+    const contentScripts = this.ast[nodeName];
+    if (!Array.isArray(contentScripts)) {
+      return;
+    }
+    for (const script of contentScripts) {
+      if (script.js) {
+        script.js = this.processMultipleDependencies(script.js);
+        this.isAstDirty = true;
+      }
+      if (script.css) {
+        script.css = this.processMultipleDependencies(script.css);
+        this.isAstDirty = true;
+      }
+    }
+  }
+
+  processWebAccessibleResources(nodeName) {
+    if (nodeName !== 'web_accessible_resources') {
+      return;
+    }
+
+    const webAccessibleResources = this.ast[nodeName];
+    if (!Array.isArray(webAccessibleResources)) {
+      return;
+    }
+    this.ast[nodeName] = this.processMultipleDependencies(
+      webAccessibleResources
+    );
+    this.isAstDirty = true;
+  }
+
+  processBrowserOrPageAction(nodeName) {
+    if (!['browser_action', 'page_action'].includes(nodeName)) {
+      return;
+    }
+
+    const action = this.ast[nodeName];
+    if (action.default_popup) {
+      action.default_popup = this.processSingleDependency(action.default_popup);
+      this.isAstDirty = true;
+    }
+    if (action.default_icon) {
+      action.default_icon = this.processSingleDependency(action.default_icon);
+      this.isAstDirty = true;
+    }
+  }
+
+  processIcons(nodeName) {
+    if (nodeName !== 'icons') {
+      return;
+    }
+
+    const icons = this.ast[nodeName];
+    for (const size of Object.keys(icons)) {
+      icons[size] = this.processSingleDependency(icons[size]);
+      this.isAstDirty = true;
+    }
+  }
+
+  collectDependenciesForWebExtension() {
+    for (const nodeName of Object.keys(this.ast)) {
```

Comment:
Love the breakout of individual concerns here. Would you consider a lookup strategy for extracting and processing the parts?

```js
const processors = {
  'background': this.proceesBackground,
  'content_scripts': this.processContentScripts,
  'web_accessible_resources': this.processWebAccessibleResources,
  // . . .
}

for (const nodeName of Object.keys(this.ast)) {
  processors[nodeName] && processors[nodeName].call(this, this.ast[nodeName])
}
```

The `process*` methods would change a little to accept the `node` from the `ast` instead of extracting it themselves:

```js
processContentScripts (contentScripts) {
  // . . . as above . . .
}
```